File: examples/automation-ruby/parallel_run.rb
# ==============================================================================
# PARALLEL RUN — reserve N devices in one group, drive each in its own thread,
# then release the group once.
#
# Usage:
#   export MERCURY_BASE_URL=https://YOUR_DOMAIN
#   export MERCURY_TOKEN=...            # From UI > Settings > Keys > Access Tokens
#   export MERCURY_TYPE=android         # android | ios
#   export MERCURY_AMOUNT=2             # how many devices? (non-admins: max 2)
#   # To pick specific devices instead:
#   # export MERCURY_SERIALS=SERIAL_A,SERIAL_B
#   ruby parallel_run.rb
#
# For single-device run: single_run.rb
# More patterns (CI matrix etc): docs/parallel-execution.md
#
# ==============================================================================
# ÇOKLU (PARALEL) KOŞUM — N cihazı tek grupta ayır, her cihaz için ayrı
# thread'de bağlan/test koştur, sonunda grubu tek seferde bırak.
#
# Çalıştırma:
#   export MERCURY_BASE_URL=https://YOUR_DOMAIN
#   export MERCURY_TOKEN=...            # UI > Settings > Keys > Access Tokens
#   export MERCURY_TYPE=android         # android | ios
#   export MERCURY_AMOUNT=2             # kaç cihaz? (admin değilsen en fazla 2)
#   # Belirli cihazlar istersen:
#   # export MERCURY_SERIALS=SERIAL_A,SERIAL_B
#   ruby parallel_run.rb
#
# Tek cihazlık koşum için: single_run.rb
# Daha fazla desen (CI matrix vb.): docs/parallel-execution.md
require_relative 'mercury_client'

client   = MercuryClient.new
serials  = ENV['MERCURY_SERIALS'].to_s.split(',').map(&:strip).reject(&:empty?)
requested_type = ENV['MERCURY_TYPE']
run_name = ENV.fetch('MERCURY_RUN', "parallel-run-#{Time.now.strftime('%Y%m%d-%H%M%S')}")

reservation = client.reserve(
  run: run_name,                                  # Run name shown on Builds page / Builds sayfasında görünen isim
  run_url: ENV['CI_JOB_URL'],                     # Optional: clickable link on Builds / Opsiyonel: Builds'de tıklanabilir link
  project: ENV['MERCURY_PROJECT'],                # Optional: groups runs on Builds / Opsiyonel: Builds'de koşumları gruplar
  timeout: Integer(ENV.fetch('MERCURY_TIMEOUT', '900')), # Seconds / Saniye
  amount: Integer(ENV.fetch('MERCURY_AMOUNT', '2')),
  type: requested_type,                           # android | ios
  serials: serials,                               # If given, reserve exactly these devices / Veriliyse tam bu cihazlar ayrılır
  need_amount: true                               # All or nothing / Hepsi ya da hiçbiri
)

group_id = reservation[:group_id]
devices  = reservation[:devices]
puts "Group #{group_id} — #{devices.length} devices: #{devices.map { |d| d['serial'] }.join(', ')}"
# Grup #{group_id} — #{devices.length} cihaz: #{devices.map { |d| d['serial'] }.join(', ')}

begin
  threads = devices.map do |device|
    Thread.new do
      serial = device['serial']
      type = client.device_type(device, requested: requested_type)
      Thread.current.name = serial
      remote = client.use_device(serial)
      puts "[#{serial}] remoteConnectUrl: #{remote}"

      if type == 'ios'
        # iOS: capability for each Appium session:
        #   'appium:webDriverAgentUrl' => remote
        # iOS: her cihaz için Appium session'ında:
        #   'appium:webDriverAgentUrl' => remote
        puts "[#{serial}] Appium capability: 'appium:webDriverAgentUrl' => '#{remote}'"
      else
        # Android: adb must run on the MACHINE where Appium is running.
        # Android: adb, Appium'un çalıştığı MAKİNEDE bağlanmalı.
        system('adb', 'connect', remote) || raise("[#{serial}] adb connect failed / başarısız")
        # Appium capability: 'appium:udid' => remote
        # Note: for parallel Appium sessions, either use different 'appium:systemPort'
        # values in a single Appium instance or run a separate Appium port per device.
        # Not: paralel Appium session'ları için ya tek Appium'da farklı
        # 'appium:systemPort' değerleri ver ya da cihaz başına ayrı Appium portu aç.
      end

      # ---- TESTS FOR THIS DEVICE RUN HERE -----------------------------------
      # ---- BU CİHAZIN TESTLERİ BURADA KOŞAR --------------------------------
      sleep Integer(ENV.fetch('MERCURY_HOLD_SECONDS', '30')) # örnek amaçlı bekleme
      # ------------------------------------------------------------------------
      puts "[#{serial}] done / tamamlandı"
    end
  end

  errors = []
  threads.each do |t|
    begin
      t.join
    rescue StandardError => e
      errors << e
      warn "[#{t.name}] ERROR / HATA: #{e.message}"
    end
  end
  raise "#{errors.length} devices had errors / cihazda hata oluştu" unless errors.empty?
ensure
  # Group released in one call; $! reports PASSED/FAILED on Builds.
  # Grup tek çağrıyla bırakılır; $! Builds'de PASSED/FAILED rozeti olur.
  client.release(group_id, result: $! ? 'failed' : 'passed')
  puts "Released group: #{group_id} / Grup bırakıldı: #{group_id}"
end

File: examples/automation-java/src/main/java/mercury/ParallelRun.java
package mercury;

import com.google.gson.JsonObject;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import static mercury.MercuryClient.env;
import static mercury.MercuryClient.envSerials;

/**
 * PARALLEL RUN — reserve N devices in one group, drive each in its own thread,
 * release the group once at the end.
 * Usage / Çalıştırma:
 *   export MERCURY_BASE_URL=... MERCURY_TOKEN=... MERCURY_TYPE=android
 *   export MERCURY_AMOUNT=2            (non-admins max 2 / admin değilsen en fazla 2)
 *   # specific devices / belirli cihazlar: export MERCURY_SERIALS=SERIAL_A,SERIAL_B
 *   mvn -q compile exec:java -Dexec.mainClass=mercury.ParallelRun
 * <p>
 * ÇOKLU (PARALEL) KOŞUM — N cihazı tek grupta ayır, her cihazı ayrı thread'de
 * koştur, sonunda grubu tek seferde bırak. Tekli için / single run: {@link SingleRun}
 */
public class ParallelRun {

    public static void main(String[] args) throws Exception {
        MercuryClient client = new MercuryClient();
        List<String> serials = envSerials();
        String runName = env("MERCURY_RUN",
            "parallel-run-" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss")));

        MercuryClient.Reservation reservation = client.reserve(
            runName,
            Integer.parseInt(env("MERCURY_TIMEOUT", "900")),
            Integer.parseInt(env("MERCURY_AMOUNT", "2")),
            env("MERCURY_TYPE", null),
            serials,                                            // if given, exactly these / verildiyse tam bunlar
            env("CI_JOB_URL", null)
        );

        String groupId = reservation.groupId();
        List<JsonObject> devices = reservation.devices();
        System.out.printf("Group %s — %d devices%n", groupId, devices.size());

        ExecutorService pool = Executors.newFixedThreadPool(devices.size());
        try {
            List<Future<?>> futures = new ArrayList<>();
            for (JsonObject device : devices) {
                futures.add(pool.submit(() -> {
                    String serial = device.get("serial").getAsString();
                    String remote = client.useDevice(serial);
                    System.out.printf("[%s] remoteConnectUrl: %s%n", serial, remote);

                    if ("ios".equals(env("MERCURY_TYPE", ""))) {
                        System.out.printf("[%s] Appium capability: appium:webDriverAgentUrl = %s%n", serial, remote);
                    }
                    else {
                        Adb.connect(remote);
                        // Note: for parallel Appium sessions use different appium:systemPort
                        // values or a separate Appium port per device.
                        // Not: paralel Appium session'ları için farklı appium:systemPort
                        // değerleri ver ya da cihaz başına ayrı Appium portu aç.
                    }

                    // ---- TESTS FOR THIS DEVICE / BU CİHAZIN TESTLERİ ------------
                    Thread.sleep(Integer.parseInt(env("MERCURY_HOLD_SECONDS", "30")) * 1000L);
                    // --------------------------------------------------------------
                    System.out.printf("[%s] done / tamamlandı%n", serial);
                    return null;
                }));
            }

            int errors = 0;
            for (Future<?> f : futures) {
                try {
                    f.get();
                }
                catch (Exception e) {
                    errors++;
                    System.err.println("ERROR / HATA: " + e.getCause());
                }
            }
            if (errors > 0) {
                throw new RuntimeException(errors + " devices had errors / cihazda hata oluştu");
            }
        }
        finally {
            pool.shutdownNow();
            client.release(groupId); // one call frees the whole group / tek çağrı tüm grubu bırakır
            System.out.println("Released group / Grup bırakıldı: " + groupId);
        }
    }
}
